### ISSUE TITLE
ArrayStore exception when calling a ProxyExecutable from JS

### ISSUE THREAD

**Issue by vmutafov** [oracle/graal#5796]
**Describe GraalVM and your environment :**
 - GraalVM version or commit id if built from source: **22.3**
 - CE or EE: **CE**
 - JDK version: **19**
 - OS and OS Version: **MacOS Monterey 12.3**
 - Architecture: **arm64 - M1 Pro**
 - The output of `java -Xinternalversion`:
```
OpenJDK 64-Bit Server VM (19.0.1+10-jvmci-22.3-b08) for bsd-aarch64 JRE (19.0.1+10-jvmci-22.3-b08), built on 2022-10-19T10:54:34Z by "graal" with clang Apple LLVM 12.0.0 (clang-1200.0.32.29)
```

**Have you verified this issue still happens when using the latest snapshot?**
No, I haven't tested with dev builds

**Describe the issue**
Initially discussed at: https://graalvm.slack.com/archives/CN9L59P61/p1673997313335259
The issue is somewhat similar to https://github.com/oracle/graaljs/issues/36 . In my case, I'm starting several Vertx verticles each having a GraalJS context with a shared engine attached to them. At the startup of each verticle I evaluate a simple JS script that calls a function bound to the GraalJS context when created. The JS script is:
```JS
proxyFun((a, b) => {
    console.log("!!! VM: a + b: " + (a + b));
   }, 1000, 2000);
```

The Kotlin function being called is:
```Kotlin
class ProxyFun(private val vertx: Vertx) : ProxyExecutable {
    override fun execute(vararg arguments: Value?): Any? {
        val fn = arguments[0]
        val args = arguments.slice(1 until arguments.size).toTypedArray()

        val task = { fn!!.executeVoid(*args) }
        vertx.runOnContext {
            task() // this crashes once or twice when running all verticles
        }
        return null
    }

}
```
and it's attached to the context:
```Kotlin
context.getBindings("js").putMember("proxyFun", ProxyFun(vertx))
```
When I run my program, it crashes in several of the verticles with an exception like the one below.
The strange part is that if I replace `vertx.runOnContext { task() }` with only `task()` it works without crashes (at least from running it 5-6 times). It also works if I initially copy all the args I pass to `fn.executeVoid`  into a new array via `System.arraycopy(arguments, 0, o, 0, o.size)` OR if I remove the engine reuse for the GraalJS contexts.

**Code snippet or code repository that reproduces the issue**
```
https://github.com/vmutafov/graaljs-arraystoreexception-reproducer
```

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. Clone the repository
2. Open it in IntelliJ/Eclipse
3. Run the `main` function in https://github.com/vmutafov/graaljs-arraystoreexception-reproducer/blob/main/src/main/java/main.kt
4. You should see in the console the exception

**Expected behavior**
Not to throw an exception or if it's a problem on my side to throw something more precise

**Additional context**

Exception:
```
org.graalvm.polyglot.PolyglotException: java.lang.ArrayStoreException: java.lang.Integer
	at com.oracle.truffle.polyglot.PolyglotLanguageContext$ToGuestValuesNode.fastToGuestValuesUnroll(PolyglotLanguageContext.java:1091)
	at com.oracle.truffle.polyglot.PolyglotLanguageContext$ToGuestValuesNode.apply(PolyglotLanguageContext.java:1061)
	at com.oracle.truffle.polyglot.PolyglotValueDispatch$InteropValue$AbstractExecuteNode.executeShared(PolyglotValueDispatch.java:4266)
	at com.oracle.truffle.polyglot.PolyglotValueDispatch$InteropValue$ExecuteVoidNode.executeImpl(PolyglotValueDispatch.java:4296)
	at com.oracle.truffle.polyglot.HostToGuestRootNode.execute(HostToGuestRootNode.java:124)
	at org.graalvm.polyglot.Value.executeVoid(Value.java:902)
	at ProxyFun$execute$crashyTask$1.invoke(main.kt:66)
	at ProxyFun$execute$crashyTask$1.invoke(main.kt:66)
	at ProxyFun.execute$lambda$0(main.kt:71)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:246)
	at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:43)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask$$$capture(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Original Internal Error: 
java.lang.ArrayStoreException: java.lang.Integer
	at com.oracle.truffle.polyglot.PolyglotLanguageContext$ToGuestValuesNode.fastToGuestValuesUnroll(PolyglotLanguageContext.java:1091)
	at com.oracle.truffle.polyglot.PolyglotLanguageContext$ToGuestValuesNode.apply(PolyglotLanguageContext.java:1061)
	at com.oracle.truffle.polyglot.PolyglotValueDispatch$InteropValue$AbstractExecuteNode.executeShared(PolyglotValueDispatch.java:4266)
	at com.oracle.truffle.polyglot.PolyglotValueDispatch$InteropValue$ExecuteVoidNode.executeImpl(PolyglotValueDispatch.java:4296)
	at com.oracle.truffle.polyglot.HostToGuestRootNode.execute(HostToGuestRootNode.java:124)
	at com.oracle.truffle.api.impl.DefaultCallTarget.callDirectOrIndirect(DefaultCallTarget.java:85)
	at com.oracle.truffle.api.impl.DefaultCallTarget.call(DefaultCallTarget.java:102)
	at com.oracle.truffle.api.impl.DefaultRuntimeAccessor$DefaultRuntimeSupport.callProfiled(DefaultRuntimeAccessor.java:182)
	at com.oracle.truffle.polyglot.PolyglotValueDispatch$InteropValue.executeVoid(PolyglotValueDispatch.java:2386)
	at org.graalvm.polyglot.Value.executeVoid(Value.java:902)
	at ProxyFun$execute$crashyTask$1.invoke(main.kt:66)
	at ProxyFun$execute$crashyTask$1.invoke(main.kt:66)
	at ProxyFun.execute$lambda$0(main.kt:71)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:246)
	at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:43)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask$$$capture(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Screenshot from the IntelliJ debugger on the crash site:
<img width="1056" alt="Screenshot 2023-01-18 at 14 29 51" src="https://user-images.githubusercontent.com/39677168/213447468-3950ffc6-5584-4b22-b03d-b359c488c179.png">

**Comment by chumer** [comment]
I reduced the problem to this reproducer:
```
public class CrashTest {

    static class ProxyFun implements ProxyExecutable {
	@Override
	public Object execute(Value... arguments) {
            Value fn = arguments[0];
            Value[] args = Arrays.copyOfRange(arguments, 1, arguments.length);
            return fn.execute((Object[])args);
        }
    }

    @Test
    public void test() throws InterruptedException, ExecutionException {
	Engine engine = Engine.create();
	ExecutorService service = Executors.newFixedThreadPool(8);
	Source s = Source.create("js", " proxyFun((a, b) => (a + b), 1000, 2000) ");
	List<Future<?>> futures = new ArrayList<>();
	for (int i = 0; i < 100; i++) {
	    futures.add(service.submit(() -> {
		try(Context context = Context
		            .newBuilder()
		            .engine(engine)
		            .build()) {
		    context.getBindings("js").putMember("proxyFun", new ProxyFun());
		    context.eval(s);
		}
	    }));
	}
	for (Future<?> future : futures) {
	    future.get();
	}
	service.shutdown();
    }
}
```

**Comment by chumer** [comment]
I could reproduce this with 22.3 and 21.3.3.1.
I could not reproduce this on master (23.0). 
This was likely fixed accidentally due to some recent changes. (the crashing node was rewritten)

**Comment by vmutafov** [comment]
Thanks a lot for the quick investigation!

**Comment by chumer** [comment]
Ok found the bug. A fix is in the pipeline to land in the next CPU for 22.3 and 21.3.

**Comment by chumer** [comment]
There is also a workaround for this problem. The problem arises when `Value.execute(Object...)` is called with an array other than Object[]. For example like this example Value[].

So instead of doing this:
```
    static class ProxyFun implements ProxyExecutable {
	@Override
	public Object execute(Value... arguments) {
            Value fn = arguments[0];
            Value[] args = Arrays.copyOfRange(arguments, 1, arguments.length);
            return fn.execute((Object[])args);
        }
    }
```

Which allocates a Value[]. You can do this:

```
static class ProxyFun implements ProxyExecutable {
    @Override
    public Object execute(Value... arguments) {
        Value fn = arguments[0];
        
        Object[] args = new Object[arguments.length - 1];
        System.arraycopy(arguments, 1, args, 0, arguments.length - 1);
        return fn.execute(args);
    }
}
```

**Comment by chumer** [comment]
Backports landed for 21.3.6 and 22.3.
